File: top-k-frequent-elements-347/top_k_frequent_elements.rb
# @param {Integer[]} nums
# @param {Integer} k
# @return {Integer[]}

class Heap
    attr_reader :heap

    def initialize
        @heap = []
    end

    def parent_index index
        (index-1)/2
    end

    def left_child_index index
        2*index+1
    end

    def right_child_index index
        2*index+2
    end

    def size
        @heap.length
    end

    def is_swap_needed?(a,b)
        @heap[a][1] > @heap[b][1]
    end

    def swap(a,b) 
        @heap[a], @heap[b] = @heap[b], @heap[a]
    end


    def peek
        @heap[0]
    end

    def bubble_up idx 
        loop do
            parent_idx = parent_index(idx)
            if idx > 0 && is_swap_needed?(parent_idx, idx)
                swap(parent_idx, idx)
                idx = parent_idx
            else
                return
            end
        end
    end

    def bubble_down idx
        loop do
            swap_idx = idx
            parent_idx = parent_index swap_idx
            left_idx = left_child_index swap_idx
            right_idx = right_child_index swap_idx

            if left_idx < size && is_swap_needed?(swap_idx, left_idx)
                swap_idx=left_idx
            end

            if right_idx < size && is_swap_needed?(swap_idx, right_idx)
                swap_idx = right_idx
            end
              
            if swap_idx != idx
                swap(idx, swap_idx)
                idx = swap_idx
            else
                return
            end
        end
    end

    def heap_push(pair)
        @heap.push(pair)
        bubble_up(size - 1)
    end
    
    def heap_pop
        return nil if size.zero?
        top = peek
        swap(0, size - 1)
        @heap.pop
        bubble_down(0) unless @heap.empty?
        top
    end
    
    def heap_push_pop(pair)
        heap_push(pair)
        heap_pop
    end

end

def top_k_frequent(nums, k)
    frequency = {}
    heap = Heap.new
    nums.each do |num|
        if frequency.has_key? num
            frequency[num] += 1
        else
            frequency[num] = 1
        end
    end

    frequency.each do |key,val|
        if heap.size < k
            heap.heap_push([key,val])
        else
            heap.heap_push_pop([key,val])
        end
    end

    heap.heap.map { |arr| arr[0] }


end

File: top-k-frequent-elements-347/top_k_frequent_elements.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number[]}
 */
var topKFrequent = function(nums, k) {
    const heap = new Heap();
    const freq = new Map();

    for(num of nums) {
        freq.set(num, (freq.get(num) || 0) + 1);
    }

    for([num, frequency] of freq) {
       // console.log(num,frequency);
        if(heap.size() < k) {
            heap.heapPush([num, frequency]);
        }else{
            heap.heapPushPop([num, frequency]);
        }
        console.log(heap.peek())
    }

    return heap.heap.map(arr => arr[0]);

};

class Heap {
    constructor() {
        this.heap = [];
    }

    parentIndex = i => Math.floor((i-1)/2);

    leftChildIndex = i => 2 * i + 1;

    rightChildIndex = i => 2 * i + 2;

    size = () => this.heap.length;

    isSwapNeed = (a,b) => this.heap[a][1] > this.heap[b][1];

    swap = (a,b) => [this.heap[a], this.heap[b]] = [this.heap[b], this.heap[a]];

    peek = () => this.heap[0];

    bubbleUp(idx) {
        while(true) {
            const parentIndex = this.parentIndex(idx);
            if(idx>0 && this.isSwapNeed(parentIndex, idx)) {
                this.swap(parentIndex, idx);
                idx = parentIndex;

            }else{
                return;
            }
            
        }
    }

    bubbleDown(idx) {

        while(true) {
            
            let swapIndex = idx;
            const leftChildIndex = this.leftChildIndex(swapIndex);
            const rightChildIndex = this.rightChildIndex(swapIndex);

            const size = this.size();

            if(leftChildIndex < size && this.isSwapNeed(swapIndex, leftChildIndex)) {
                swapIndex = leftChildIndex;
            }

            if(rightChildIndex < size && this.isSwapNeed(swapIndex, rightChildIndex)) {
                swapIndex = rightChildIndex;
            }

            if(swapIndex != idx) {
                this.swap(idx, swapIndex);
                idx = swapIndex;
            }else {
                return;
            }
        }

    }

    heapPush(pair) {
        
        this.heap.push(pair);
        const idx = this.size()-1;
        this.bubbleUp(idx);

    }

    heapPop() {
        const top = this.peek();
        this.swap(0, this.size()-1);
        this.heap.pop();
        this.bubbleDown(0);
        return top;

    }

    heapPushPop(pair) {
        this.heapPush(pair);
        this.heapPop();
    }
        
}